Realm/RLMManagedDictionary.mm

```diff
@@ -0,0 +1,665 @@
+    return [self objectForKey:key];
+}
+
+- (nonnull id)objectAtIndex:(NSUInteger)index {
+    return translateErrors([&] {
+        auto key = _backingCollection.get_pair(index).first;
+        return RLMStringDataToNSString(key);
+    });
+}
+
+- (NSUInteger)indexOfObject:(id)value {
+    return translateErrors([&] {
+        return _backingCollection.find_any(RLMObjcToMixed(value));
+    });
+}
+
+- (void)setObject:(id)obj forKey:(id<RLMDictionaryKey>)key {
+    changeDictionary(self, ^{
+        RLMDictionaryValidateMatchingObjectType(self, key, obj);
+        RLMAccessorContext context(*_objectInfo);
+        _backingCollection.insert(context,
+                                  context.unbox<realm::StringData>(key),
+                                  obj);
+    });
+}
+
+- (void)setObject:(id)obj forKeyedSubscript:(id<RLMDictionaryKey>)key {
+    // passing `nil` to the subscript should delete the object.
+    if (!obj) {
+        [self removeObjectForKey:key];
+        return;
+    }
+    changeDictionary(self, ^{
+        RLMDictionaryValidateMatchingObjectType(self, key, obj);
+        RLMAccessorContext context(*_objectInfo);
+        _backingCollection.insert(context,
+                                  context.unbox<realm::StringData>(key),
+                                  obj);
+    });
+}
+
+- (void)removeAllObjects {
+    changeDictionary(self, ^{
+        _backingCollection.remove_all();
+    });
+}
+
+- (void)removeObjectsForKeys:(NSArray *)keyArray {
+    for (id key in keyArray) {
+        [self removeObjectForKey:key];
+    }
+}
+
+- (void)removeObjectForKey:(id<RLMDictionaryKey>)key {
+    try {
+        changeDictionary(self, ^{
+            RLMAccessorContext context(*_objectInfo);
+            _backingCollection.erase(context.unbox<realm::StringData>(key));
+        });
+    }
+    catch (realm::KeyNotFound const&) {
+        return;
+    }
+    catch (...) {
+        throwError(nil, nil);
+    }
+}
+
+- (void)enumerateKeysAndObjectsUsingBlock:(void (^)(id <RLMDictionaryKey> key,
+                                                    id obj, BOOL *stop))block {
+    for (id key in [self allKeys]) {
+        BOOL stop = false;
+        block(key, self[key], &stop);
+        if (stop) {
+            break;
+        }
+    }
+}
+
+#pragma mark - KVC
+
+- (id)valueForKeyPath:(NSString *)keyPath {
+    if ([keyPath hasPrefix:@"@"]) {
+        // Delegate KVC collection operators to RLMResults
+        return translateErrors([&] {
+            auto results = [RLMResults resultsWithObjectInfo:*_objectInfo
+                                                     results:_backingCollection.as_results()];
+            return [results valueForKeyPath:keyPath];
+        });
+    }
+    return [super valueForKeyPath:keyPath];
+}
+
+- (id)valueForKey:(NSString *)key {
+    if ([key hasPrefix:@"@"]) {
+        if ([key isEqualToString:[NSString stringWithFormat:@"@%@", RLMInvalidatedKey]]) {
```

Comment:
This is an awfully roundabout way of writing @"invalidated"

Reply:
I was trying to copy the behaviour of NSDictionary here.

If key does not start with “@”, invokes objectForKey:. If key does start with “@”, strips the “@” and invokes [super valueForKey:] with the rest of the key..

So without the @ prefix the key is assumed to be a value in the collection, and it might be possible someone wants to use invalidated as a key for their value.